apparmor/probe.go

```diff
@@ -0,0 +1,118 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package apparmor
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// FeatureLevel encodes the kind of support for apparmor found on this system.
+type FeatureLevel int
+
+const (
+	// None indicates that apparmor is not enabled.
+	None FeatureLevel = iota
+	// Partial indicates that apparmor is enabled but some features are missing.
+	Partial
+	// Full indicates that all features are supported.
+	Full
+)
+
+var (
+	// featureSysPath points to the sysfs directory where apparmor features are listed.
+	featuresSysPath = "/sys/kernel/security/apparmor/features"
+	// requiredFeatures are the apparmor features needed for strict confinement.
+	requiredFeatures = []string{
+		"caps",
+		"dbus",
+		"domain",
+		"file",
+		"mount",
+		"namespaces",
+		"network",
+		"ptrace",
+		"rlimit",
+		"signal",
+	}
+)
+
+// Probe checks which apparmor features are available.
+//
+// The error is returned whenever less-than-full support is detected.
+func Probe() (FeatureLevel, error) {
+	_, err := os.Stat(featuresSysPath)
+	if err != nil {
```

Comment:
This feels strange from an API perspective. One could argue that there is no error here, the fact that there is no apparmor dir is not an error, its what is expected when there is no apparmor support. The api is also strange in that the FeatureLevel is actually valid even when an error is returned (which is not common at all in go code).

Maybe FeatureLevel can become a richer type? Something like:
```
type FeatureLevel struct {
   Summary string
}
type Full struct {
   FeatureLevel
}
type Partial struct {
    Featurelevel
}
type None struct {
   FeatureLevel
}
return Partial{Summary: fmt.Sprintf("apparmor features missing: %s", strings.Join(missing, ", "))
```
or similar? And only returning an error if there is an actual error condition (or not returning one at all as it seems we have no real errors right now in the code)?

Reply:
I agree about the error comment and I'll see what I can do to make this nicer. I really like the simplicity of the enum approach so far and would prefer not to cross YAGNI until I actually need it.

Reply:
I'm working on something nice, I'll push it shortly.

Reply:
Done! I re-factored the code a little and I think you will be happy with the result.